sysl2/sysl/validate.go

```diff
@@ -0,0 +1,442 @@
+			break
+		} else {
+			if _, exists := implAttrs[name]; !exists {
+				matching = false
+			} else {
+				matching = true
+				delete(implAttrNames, name)
+				break
+			}
+		}
+	}
+
+	if !matching {
+		var implAttrNames []string
+		for k := range implAttrs {
+			implAttrNames = append(implAttrNames, k)
+		}
+		v.messages = append(v.messages,
+			*NewMsg(ErrInvalidOption, []string{viewName, strings.Join(implAttrNames, ","), specTupleName}))
+	}
+}
+
+func (v *Validator) GetMessages() []Msg {
+	return v.messages
+}
+
+func NewValidator(grammar *sysl.Application, transform *sysl.Application) *Validator {
+	return &Validator{grammar: grammar, transform: transform}
+}
+
+func (r *Resolver) ResolveAllViews() {
+	for viewName, tfmView := range r.transform.Views {
+		typeName := getTypeName(tfmView.GetRetType())
+		r.resolveExprType(tfmView.GetExpr(), viewName, viewName, typeName)
+	}
+}
+
+func (r *Resolver) resolveExprType(expr *sysl.Expr, viewName string, scope string, typeName string) *sysl.Type {
+	switch e := expr.Expr.(type) {
+	case *sysl.Expr_Transform_:
+
+		newType := &sysl.Type{
+			Type: &sysl.Type_Tuple_{
+				Tuple: &sysl.Type_Tuple{
+					AttrDefs: map[string]*sysl.Type{},
+				},
+			},
+		}
+
+		attrDefs := newType.GetTuple().AttrDefs
+
+		for _, stmt := range e.Transform.GetStmt() {
+			switch s := stmt.Stmt.(type) {
+			case *sysl.Expr_Transform_Stmt_Assign_:
+				varName := s.Assign.GetName()
+
+				expr := s.Assign.GetExpr()
+				exprType := r.resolveExprType(expr, viewName, scope+":"+varName, typeName)
+				attrDefs[varName] = exprType
+				r.assignTypes[viewName] = map[string]*sysl.Type{typeName: newType}
+			case *sysl.Expr_Transform_Stmt_Let:
+				varName := s.Let.GetName()
+				expr := s.Let.GetExpr()
+				exprType := r.resolveExprType(expr, viewName, scope+":"+varName, typeName)
+				r.varTypes[scope+":"+varName] = map[string]*sysl.Type{getTypeName(s.Let.GetExpr().GetType()): exprType}
+			}
+		}
+		return newType
+	case *sysl.Expr_Literal:
+		return expr.GetType()
+	case *sysl.Expr_Unexpr:
+		varType := r.resolveExprType(expr.GetUnexpr().GetArg(), viewName, scope, typeName)
+		switch e.Unexpr.GetOp() {
+		case sysl.Expr_UnExpr_NOT, sysl.Expr_UnExpr_INV:
+			if !hasSameType(varType, typeBool()) {
+				_, typeDetail := getTypeDetail(varType)
+				r.messages = append(r.messages, *NewMsg(ErrInvalidUnary, []string{viewName, typeDetail}))
+			}
+
+			return typeBool()
+		case sysl.Expr_UnExpr_NEG, sysl.Expr_UnExpr_POS:
+			if !hasSameType(varType, typeInt()) {
+				_, typeDetail := getTypeDetail(varType)
+				r.messages = append(r.messages, *NewMsg(ErrInvalidUnary, []string{viewName, typeDetail}))
+			}
+
+			return typeInt()
+		}
+	}
+
+	return typeNone()
+}
+
+func (r *Resolver) getResolvedAssigns() map[string]map[string]*sysl.Type {
+	return r.assignTypes
+}
+
+func (r *Resolver) getResolvedVars() map[string]map[string]*sysl.Type {
```

Comment:
`map[string]map[string]*sysl.Type` looks weird.
the second map is just Type_Tuple.

Reply:
This is to preserver details from view.

First map contains tuples by view name.
Second map contains tuple by tuple name.

eg:
```
!view goFileView(app <: sysl.App) -> goFile:  
  app -> (:  
    TopLevelDecl = app -> <TopLevelDecl> (:  
      Foo = ""  
    )  
  )
```


```
"goFileView":
      "goFile": 
           Tuple{
               AttrDefs: 
                   "TopLevelDecl":
                              Tuple{
		                  AttrDefs: "Foo": Type_Primitive
			     }
	       }
```